src/FeedGenerator.php

```diff
@@ -629,39 +710,36 @@ public function handle_action_timeout( $action_id, $error ) {
 	 * @since 1.2.14
 	 */
 	protected function is_timeout_error( array $error ): bool {
-		return isset( $error['type'] ) &&
-				E_ERROR === $error['type'] &&
-				isset( $error['message'] ) &&
-				strpos( $error ['message'], 'Maximum execution time' ) !== false;
+		$is_error              = E_ERROR === $error['type'] ?? 0;
+		$is_max_execution_time = strpos( $error['message'] ?? '', 'Maximum execution time' ) !== false;
+		return $is_error && $is_max_execution_time;
 	}
 
 	/**
-	 * Handle error on generate feed timeout.
+	 * Check whether the action's failure rate is above the specified threshold within the timeframe.
 	 *
-	 * @since 1.2.14
+	 * @param string $hook The job action hook.
+	 * @param ?array $args The job arguments.
 	 *
-	 * @param int $action_id The ID of the action marked as failed.
+	 * @return bool True if the action's error rate is above the threshold, and false otherwise.
 	 *
-	 * @throws Exception Related to max retries reached or missing arguments on the action.
+	 * @since x.x.x
 	 */
-	public function maybe_handle_error_on_timeout( int $action_id ) {
-
-		$action = ActionScheduler::store()->fetch_action( $action_id );
-
-		if ( $this->get_action_full_name( self::CHAIN_BATCH ) !== $action->get_hook() ) {
-			return;
-		}
-
-		try {
-			$action_args = $action->get_args();
-
-			if ( ! isset( $action_args[0] ) || ! is_int( $action_args[0] ) || ! isset( $action_args[1] ) || ! is_array( $action_args[1] ) ) {
-				throw new Exception( __( 'There was not possible to re-schedule the action, no args available.', 'pinterest-for-woocommerce' ) );
-			}
+	protected function is_failure_rate_above_threshold( string $hook, ?array $args = null ): bool {
+		$threshold   = apply_filters( 'pinterest_for_woocommerce_action_failure_threshold', 3 );
+		$time_period = apply_filters( 'pinterest_for_woocommerce_action_failure_time_period', 1800 ); // 30 minutes.
```

Comment:
Minor: Let's change this to `30 * MINUTE_IN_SECONDS`, to make it more obvious what the number is without the need for the trailing comment.

```suggestion
		$time_period = apply_filters( 'pinterest_for_woocommerce_action_failure_time_period', 30 * MINUTE_IN_SECONDS );
```